### ISSUE TITLE
Error running on centos7

### ISSUE THREAD

**Issue by jonesmz** [iovisor/bcc#1530]
I'm trying to get bcc (specifically the memleak checker) working on centos7.

I've put together a modified rpm package spec, based on https://repo.iovisor.org/yum/nightly/f27/SRPMS/:

```
%bcond_with local_clang_static
#lua jit not available for some architectures
%ifarch ppc64 aarch64 ppc64le
%{!?with_lua: %global with_lua 0}
%else
%{!?with_lua: %global with_lua 1}
%endif
%define debug_package %{nil}

%define llvmver 4.0.1

Name:           bcc
Version:        0.5.0
Release:        79.git.3763380
Summary:        BPF Compiler Collection (BCC)

Group:          Development/Languages
License:        ASL 2.0
URL:            https://github.com/iovisor/bcc
Source0:        bcc-0.5.0.tar.gz
Source1:        http://llvm.org/releases/%{llvmver}/llvm-%{llvmver}.src.tar.xz
Source2:        http://llvm.org/releases/%{llvmver}/cfe-%{llvmver}.src.tar.xz

ExclusiveArch: x86_64 ppc64 aarch64 ppc64le
BuildRequires: bison cmake >= 2.8.7 flex make
BuildRequires: gcc gcc-c++ python2-devel elfutils-libelf-devel-static
%if %{with_lua}
BuildRequires: luajit luajit-devel
%endif
#%if %{without local_clang_static}
#BuildRequires: llvm-devel llvm-static
#BuildRequires: clang-devel
#%endif
BuildRequires: pkgconfig ncurses-devel

%description
Python bindings for BPF Compiler Collection (BCC). Control a BPF program from
userspace.

%if %{with_lua}
%global lua_include `pkg-config --variable=includedir luajit`
%global lua_libs `pkg-config --variable=libdir luajit`/lib`pkg-config --variable=libname luajit`.so
%global lua_config -DLUAJIT_INCLUDE_DIR=%{lua_include} -DLUAJIT_LIBRARIES=%{lua_libs}
%endif

%prep
%setup -T -b 1 -n llvm-%{llvmver}.src
mkdir -p tools/clang
tar -xvvJf %{_sourcedir}/cfe-%{llvmver}.src.tar.xz -C tools/clang --strip 1
%setup -D -n bcc

%build

export LD_LIBRARY_PATH="%{_builddir}/usr/lib64"
export PATH="%{_builddir}/usr/bin":$PATH

# build llvm
pushd %{_builddir}/llvm-%{llvmver}.src
mkdir -p build
cd build
cmake3 .. -DCMAKE_INSTALL_PREFIX=%{_builddir}/usr -DLLVM_ENABLE_ASSERTIONS:BOOL=false -DCMAKE_BUILD_TYPE:STRING=Release
make %{?_smp_mflags}
make install
popd

mkdir -p build
pushd build
cmake3 .. -DREVISION_LAST=%{version} -DREVISION=%{version} -DCMAKE_INSTALL_PREFIX=/usr \
      -DCMAKE_CXX_STANDARD=11 -DCMAKE_CXX_STANDARD_REQUIRED=ON -DCXX_STANDARD=11 \
      -DCMAKE_RULE_MESSAGES:BOOL=OFF -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON \
      %{?lua_config}
make %{?_smp_mflags}
popd

%install
pushd build
make install/strip DESTDIR=%{buildroot}

%package -n libbcc
Summary: Shared Library for BPF Compiler Collection (BCC)
Requires: elfutils-libelf
%description -n libbcc
Shared Library for BPF Compiler Collection (BCC)

%package -n python-bcc
Summary: Python bindings for BPF Compiler Collection (BCC)
Requires: libbcc = %{version}-%{release}
%description -n python-bcc
Python bindings for BPF Compiler Collection (BCC)

%if %{with_lua}
%package -n bcc-lua
Summary: Standalone tool to run BCC tracers written in Lua
Requires: libbcc = %{version}-%{release}
%description -n bcc-lua
Standalone tool to run BCC tracers written in Lua
%endif

%package -n libbcc-examples
Summary: Examples for BPF Compiler Collection (BCC)
Requires: python-bcc = %{version}-%{release}
%if %{with_lua}
Requires: bcc-lua = %{version}-%{release}
%endif
%description -n libbcc-examples
Examples for BPF Compiler Collection (BCC)

%package -n bcc-tools
Summary: Command line tools for BPF Compiler Collection (BCC)
Requires: python-bcc = %{version}-%{release}
%description -n bcc-tools
Command line tools for BPF Compiler Collection (BCC)

%files -n libbcc
/usr/lib64/*
/usr/include/bcc/*

%files -n python-bcc
%{python_sitelib}/bcc*

%if %{with_lua}
%files -n bcc-lua
/usr/bin/bcc-lua
%endif

%files -n libbcc-examples
/usr/share/bcc/examples/*
%exclude /usr/share/bcc/examples/*.pyc
%exclude /usr/share/bcc/examples/*.pyo
%exclude /usr/share/bcc/examples/*/*.pyc
%exclude /usr/share/bcc/examples/*/*.pyo
%exclude /usr/share/bcc/examples/*/*/*.pyc
%exclude /usr/share/bcc/examples/*/*/*.pyo

%files -n bcc-tools
/usr/share/bcc/introspection/*
/usr/share/bcc/tools/*
/usr/share/bcc/man/*

%post -n libbcc -p /sbin/ldconfig

%postun -n libbcc -p /sbin/ldconfig

%changelog
* Mon Nov 21 2016 William Cohen <wcohen@redhat.com> - 0.2.0-1
- Revise bcc.spec to address rpmlint issues and build properly in Fedora koji.

* Mon Apr 04 2016 Vicent Marti <vicent@github.com> - 0.1.4-1
- Add bcc-lua package

* Sun Nov 29 2015 Brenden Blanco <bblanco@plumgrid.com> - 0.1.3-1
- Add bcc-tools package

* Mon Oct 12 2015 Brenden Blanco <bblanco@plumgrid.com> - 0.1.2-1
- Add better version numbering into libbcc.so

* Fri Jul 03 2015 Brenden Blanco <bblanco@plumgrid.com> - 0.1.1-2
- Initial RPM Release
```

I get this error when running.

```
admin@localhost:~/rpmbuild/RPMS/x86_64 $ /usr/share/bcc/tools/memleak
/virtual/main.c:202:1: warning: declaration of 'struct tracepoint__kmem__kmalloc' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, kmalloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmalloc
^
/virtual/main.c:203:53: error: incomplete definition of type 'struct tracepoint__kmem__kmalloc'
        gen_alloc_enter((struct pt_regs *)args, args->bytes_alloc);
                                                ~~~~^
/virtual/main.c:202:1: note: forward declaration of 'struct tracepoint__kmem__kmalloc'
TRACEPOINT_PROBE(kmem, kmalloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmalloc
^
/virtual/main.c:204:68: error: incomplete definition of type 'struct tracepoint__kmem__kmalloc'
        return gen_alloc_exit2((struct pt_regs *)args, (size_t)args->ptr);
                                                               ~~~~^
/virtual/main.c:202:1: note: forward declaration of 'struct tracepoint__kmem__kmalloc'
TRACEPOINT_PROBE(kmem, kmalloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmalloc
^
/virtual/main.c:207:1: warning: declaration of 'struct tracepoint__kmem__kmalloc_node' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, kmalloc_node) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmalloc_node
^
/virtual/main.c:208:53: error: incomplete definition of type 'struct tracepoint__kmem__kmalloc_node'
        gen_alloc_enter((struct pt_regs *)args, args->bytes_alloc);
                                                ~~~~^
/virtual/main.c:207:1: note: forward declaration of 'struct tracepoint__kmem__kmalloc_node'
TRACEPOINT_PROBE(kmem, kmalloc_node) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmalloc_node
^
/virtual/main.c:209:68: error: incomplete definition of type 'struct tracepoint__kmem__kmalloc_node'
        return gen_alloc_exit2((struct pt_regs *)args, (size_t)args->ptr);
                                                               ~~~~^
/virtual/main.c:207:1: note: forward declaration of 'struct tracepoint__kmem__kmalloc_node'
TRACEPOINT_PROBE(kmem, kmalloc_node) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmalloc_node
^
/virtual/main.c:212:1: warning: declaration of 'struct tracepoint__kmem__kfree' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, kfree) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kfree
^
/virtual/main.c:213:67: error: incomplete definition of type 'struct tracepoint__kmem__kfree'
        return gen_free_enter((struct pt_regs *)args, (void *)args->ptr);
                                                              ~~~~^
/virtual/main.c:212:1: note: forward declaration of 'struct tracepoint__kmem__kfree'
TRACEPOINT_PROBE(kmem, kfree) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kfree
^
/virtual/main.c:216:1: warning: declaration of 'struct tracepoint__kmem__kmem_cache_alloc' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, kmem_cache_alloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_alloc
^
/virtual/main.c:217:53: error: incomplete definition of type 'struct tracepoint__kmem__kmem_cache_alloc'
        gen_alloc_enter((struct pt_regs *)args, args->bytes_alloc);
                                                ~~~~^
/virtual/main.c:216:1: note: forward declaration of 'struct tracepoint__kmem__kmem_cache_alloc'
TRACEPOINT_PROBE(kmem, kmem_cache_alloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_alloc
^
/virtual/main.c:218:68: error: incomplete definition of type 'struct tracepoint__kmem__kmem_cache_alloc'
        return gen_alloc_exit2((struct pt_regs *)args, (size_t)args->ptr);
                                                               ~~~~^
/virtual/main.c:216:1: note: forward declaration of 'struct tracepoint__kmem__kmem_cache_alloc'
TRACEPOINT_PROBE(kmem, kmem_cache_alloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_alloc
^
/virtual/main.c:221:1: warning: declaration of 'struct tracepoint__kmem__kmem_cache_alloc_node' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, kmem_cache_alloc_node) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_alloc_node
^
/virtual/main.c:222:53: error: incomplete definition of type 'struct tracepoint__kmem__kmem_cache_alloc_node'
        gen_alloc_enter((struct pt_regs *)args, args->bytes_alloc);
                                                ~~~~^
/virtual/main.c:221:1: note: forward declaration of 'struct tracepoint__kmem__kmem_cache_alloc_node'
TRACEPOINT_PROBE(kmem, kmem_cache_alloc_node) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_alloc_node
^
/virtual/main.c:223:68: error: incomplete definition of type 'struct tracepoint__kmem__kmem_cache_alloc_node'
        return gen_alloc_exit2((struct pt_regs *)args, (size_t)args->ptr);
                                                               ~~~~^
/virtual/main.c:221:1: note: forward declaration of 'struct tracepoint__kmem__kmem_cache_alloc_node'
TRACEPOINT_PROBE(kmem, kmem_cache_alloc_node) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_alloc_node
^
/virtual/main.c:226:1: warning: declaration of 'struct tracepoint__kmem__kmem_cache_free' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, kmem_cache_free) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_free
^
/virtual/main.c:227:67: error: incomplete definition of type 'struct tracepoint__kmem__kmem_cache_free'
        return gen_free_enter((struct pt_regs *)args, (void *)args->ptr);
                                                              ~~~~^
/virtual/main.c:226:1: note: forward declaration of 'struct tracepoint__kmem__kmem_cache_free'
TRACEPOINT_PROBE(kmem, kmem_cache_free) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__kmem_cache_free
^
/virtual/main.c:230:1: warning: declaration of 'struct tracepoint__kmem__mm_page_alloc' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, mm_page_alloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__mm_page_alloc
^
/virtual/main.c:231:61: error: incomplete definition of type 'struct tracepoint__kmem__mm_page_alloc'
        gen_alloc_enter((struct pt_regs *)args, 4096 << args->order);
                                                        ~~~~^
/virtual/main.c:230:1: note: forward declaration of 'struct tracepoint__kmem__mm_page_alloc'
TRACEPOINT_PROBE(kmem, mm_page_alloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__mm_page_alloc
^
/virtual/main.c:232:60: error: incomplete definition of type 'struct tracepoint__kmem__mm_page_alloc'
        return gen_alloc_exit2((struct pt_regs *)args, args->pfn);
                                                       ~~~~^
/virtual/main.c:230:1: note: forward declaration of 'struct tracepoint__kmem__mm_page_alloc'
TRACEPOINT_PROBE(kmem, mm_page_alloc) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__mm_page_alloc
^
/virtual/main.c:235:1: warning: declaration of 'struct tracepoint__kmem__mm_page_free' will not be visible outside of this function [-Wvisibility]
TRACEPOINT_PROBE(kmem, mm_page_free) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__mm_page_free
^
/virtual/main.c:236:67: error: incomplete definition of type 'struct tracepoint__kmem__mm_page_free'
        return gen_free_enter((struct pt_regs *)args, (void *)args->pfn);
                                                              ~~~~^
/virtual/main.c:235:1: note: forward declaration of 'struct tracepoint__kmem__mm_page_free'
TRACEPOINT_PROBE(kmem, mm_page_free) {
^
/virtual/include/bcc/helpers.h:597:46: note: expanded from macro 'TRACEPOINT_PROBE'
int tracepoint__##category##__##event(struct tracepoint__##category##__##event *args)
                                             ^
<scratch space>:81:1: note: expanded from here
tracepoint__kmem__mm_page_free
^
/virtual/main.c:15:1: error: could not open bpf map: Operation not permitted
is maps/hash map type enabled in your kernel?
BPF_HASH(sizes, u64);
^
/virtual/include/bcc/helpers.h:120:48: note: expanded from macro 'BPF_HASH'
  BPF_HASHX(__VA_ARGS__, BPF_HASH4, BPF_HASH3, BPF_HASH2, BPF_HASH1)(__VA_ARGS__)
                                               ^
/virtual/main.c:16:1: error: could not open bpf map: Operation not permitted
is maps/hash map type enabled in your kernel?
BPF_TABLE("hash", u64, struct alloc_info_t, allocs, 1000000);
^
/virtual/include/bcc/helpers.h:59:76: note: expanded from macro 'BPF_TABLE'
#define BPF_TABLE(_table_type, _key_type, _leaf_type, _name, _max_entries) \
                                                                           ^
/virtual/include/bcc/helpers.h:55:4: note: expanded from macro '\
BPF_F_TABLE'
}; \
   ^
/virtual/main.c:17:1: error: could not open bpf map: Operation not permitted
is maps/hash map type enabled in your kernel?
BPF_HASH(memptrs, u64, u64);
^
/virtual/include/bcc/helpers.h:120:37: note: expanded from macro 'BPF_HASH'
  BPF_HASHX(__VA_ARGS__, BPF_HASH4, BPF_HASH3, BPF_HASH2, BPF_HASH1)(__VA_ARGS__)
                                    ^
/virtual/main.c:18:1: error: could not open bpf map: Operation not permitted
is maps/stacktrace map type enabled in your kernel?
BPF_STACK_TRACE(stack_traces, 10240)
^
/virtual/include/bcc/helpers.h:187:3: note: expanded from macro 'BPF_STACK_TRACE'
  BPF_TABLE("stacktrace", int, struct bpf_stacktrace, _name, _max_entries);
  ^
/virtual/include/bcc/helpers.h:59:76: note: expanded from macro 'BPF_TABLE'
#define BPF_TABLE(_table_type, _key_type, _leaf_type, _name, _max_entries) \
                                                                           ^
/virtual/include/bcc/helpers.h:55:4: note: expanded from macro '\
BPF_F_TABLE'
}; \
   ^
/virtual/main.c:19:1: error: could not open bpf map: Operation not permitted
is maps/hash map type enabled in your kernel?
BPF_TABLE("hash", u64, struct combined_alloc_info_t, combined_allocs, 10240);
^
/virtual/include/bcc/helpers.h:59:76: note: expanded from macro 'BPF_TABLE'
#define BPF_TABLE(_table_type, _key_type, _leaf_type, _name, _max_entries) \
                                                                           ^
/virtual/include/bcc/helpers.h:55:4: note: expanded from macro '\
BPF_F_TABLE'
}; \
   ^
/virtual/main.c:25:26: error: bpf_table combined_allocs failed to open
        existing_cinfo = combined_allocs.lookup(&stack_id);
                         ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
8 warnings and 20 errors generated.
Traceback (most recent call last):
  File "/usr/share/bcc/tools/memleak", line 387, in <module>
    bpf = BPF(text=bpf_source)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 301, in __init__
    raise Exception("Failed to compile BPF module %s" % src_file)
Exception: Failed to compile BPF module
```

Is this an issue with the way I've compiled bcc?

Or perhaps a problem with my kernel headers?

Something else?

**Comment by yonghong-song** [comment]
The first error is:
```

/virtual/main.c:203:53: error: incomplete definition of type 'struct
tracepoint__kmem__kmalloc'
        gen_alloc_enter((struct pt_regs *)args, args->bytes_alloc);
```

This means that bcc is not able to access
/sys/kernel/debug/tracing/events/kmem/kmalloc tracepoint.

Could you double check your kernel config has CONFIG_TRACING,
CONFIG_EVENT_TRACING on?

There are also a few errors like
```

/virtual/main.c:15:1: error: could not open bpf map: Operation not permitted
is maps/hash map type enabled in your kernel?
BPF_HASH(sizes, u64);
```

Could you check whether you have necessary BPF related kernel configs as

in https://github.com/iovisor/bcc/blob/master/INSTALL.md?

Thanks!

**Comment by jonesmz** [comment]
Thank you. I had the wrong kernel .config file. Swapping it and re-running make for the kernel (to get autoconf.h) fixed the problem.

I have a new issue here: https://github.com/iovisor/bcc/issues/1535
